### PR TITLE
(closes #879) Rename GlobalInterface to ImportInterface

### LINKED ISSUE #879: Rename GlobalInterface
When defining the interface of a Symbol (which describe where the symbol referenced data comes from) we have a GlobalInterface to specify that the data storage is not local and the lifetime is longer than the current scope. This is a confusing name as symbols already have the Global concept regarding its symbol scope (given by in which symbol table the symbol is declared in the PSyIR tree).

Initially we intended this to be a generic class for several non-local storage classes (see #401) but maybe we should make a different interface for each of them and rename the current `GlobalInteraface(container_name)` that we use to import symbols from other modules as `ExternalInterface(container_name)`  or even `ExternalContainerInterface(container_name)`.

### PR DESCRIPTION
**Pull request by sergisiso**
Global interface was a bad name as since generally what is a global symbol depends on the scope. A better name for the interface (how the data comes into the scope) is External or Import Interface. I decided for the later.

Before we could have some confusing Local scope symbols with global interface (now local scope imports) and global scope symbols with global interface (now global scope imports).

### CODE REVIEW

**Comment by sergisiso** [comment]
@rupertford @arporter This is ready for review, it touches many files but it is essentially just a rename of the Interface plus the rename of a few related concepts (symbol table methods, a transformation) and I added a bit of documentation about the Symbol interfaces in the User Guide.

File: src/psyclone/psyGen.py [hunk 1]
```diff
@@ -1021,15 +1021,15 @@ def gen_code(self, parent):
         self.parent._symbol_table = self.parent.symbol_table.shallow_copy()
         # pylint: enable=protected-access
 
-        # Global symbols promoted from Kernel Globals are in the SymbolTable
-        # First aggregate all globals variables from the same module in a map
+        # Imported symbols promoted from Kernel imports are in the SymbolTable.
+        # First aggregate all imports variables from the same module in a map.
```
Comment (arporter) [inline src/psyclone/psyGen.py:1025]: "...all variables imported from the same module in a map."

File: src/psyclone/psyir/backend/fortran.py [hunk 2]
```diff
@@ -680,7 +680,7 @@ def gen_decls(self, symbol_table, args_allowed=True):
         :rtype: str
 
         :raises VisitorError: if one of the symbols is a RoutineSymbol \
-            which does not have a GlobalInterface or LocalInterface as this \
+            which does not have a ImportInterface or LocalInterface as this \
```
Comment (arporter) [inline src/psyclone/psyir/backend/fortran.py:683]: s/have a ImportInterface/have an ImportInterface/

File: src/psyclone/psyir/backend/fortran.py [hunk 3]
```diff
@@ -706,7 +706,7 @@ def gen_decls(self, symbol_table, args_allowed=True):
                     wildcard_imports_checked = True
                 if not has_wildcard_import:
                     raise VisitorError(
-                        "Routine symbol '{0}' does not have a GlobalInterface "
+                        "Routine symbol '{0}' does not have a ImportInterface "
```
Comment (arporter) [inline src/psyclone/psyir/backend/fortran.py:709]: s/a/an/

File: src/psyclone/psyir/nodes/psy_data_node.py [hunk 4]
```diff
@@ -137,8 +137,7 @@ def __init__(self, ast=None, children=None, parent=None, options=None):
         # name of the PSyData derived type) but we keep a list for future
         # extensibility.
         _PSyDataSymbol = namedtuple("_PSyDataSymbol", "name symbol_type")
-        self.imported_symbols = [_PSyDataSymbol(self.type_name,
-                                                DataTypeSymbol)]
+        self.import_symbols = [_PSyDataSymbol(self.type_name, DataTypeSymbol)]
```
Comment (arporter) [inline src/psyclone/psyir/nodes/psy_data_node.py:140]: A small point but I prefer the original `imported_symbols` - any reason for the change?

File: src/psyclone/psyir/symbols/symbol.py [hunk 5]
```diff
@@ -97,11 +96,11 @@ def __init__(self, container_symbol):
         # pylint: disable=import-outside-toplevel
         from psyclone.psyir.symbols import ContainerSymbol
 
-        super(GlobalInterface, self).__init__()
+        super(ImportInterface, self).__init__()
 
         if not isinstance(container_symbol, ContainerSymbol):
             raise TypeError(
-                "Global container_symbol parameter must be of type"
+                "Import container_symbol parameter must be of type"
```
Comment (arporter) [inline src/psyclone/psyir/symbols/symbol.py:103]: s/Import/ImportInterface/ will be clearer I think.

File: src/psyclone/psyir/symbols/symbol.py [hunk 6]
```diff
@@ -303,16 +302,16 @@ def specialise(self, subclass):
...
         '''
-        Looks-up and returns the Symbol referred to by this Symbol's external
-        (global) interface.
+        Looks-up and returns the Symbol referred to by this Symbol's
+        Import Interface.
 
         :raises SymbolError: if the module pointed to by the symbol interface \
                              does not contain the symbol (or the symbol is \
                              not public).
         :raises NotImplementedError: if the this symbol does not have an \
```
Comment (arporter) [inline src/psyclone/psyir/symbols/symbol.py:311]: s/the this/this/

File: src/psyclone/psyir/symbols/symbol.py [hunk 7]
```diff
@@ -341,15 +340,15 @@ def resolve_deferred(self):
         class and type of the looked-up symbol are the same as this one,
         some specialisations of this method update the differing
         properties in place rather than create a new symbol.
-        If this symbol does not have a 'global' interface then there is no
+        If this symbol does not have a ImportInterface then there is no
```
Comment (arporter) [inline src/psyclone/psyir/symbols/symbol.py:343]: "...have an ImportInterface..."

File: src/psyclone/psyir/symbols/symbol.py [hunk 8]
```diff
@@ -428,13 +427,13 @@ def is_local(self):
         return isinstance(self._interface, LocalInterface)
 
     @property
-    def is_global(self):
+    def is_import(self):
         '''
-        :returns: whether the Symbol has a Global interface.
+        :returns: whether the Symbol has a Import interface.
```
Comment (arporter) [inline src/psyclone/psyir/symbols/symbol.py:432]: s/a Import/an Import/

File: src/psyclone/psyir/symbols/symboltable.py [hunk 9]
```diff
@@ -903,14 +903,14 @@ def argument_datasymbols(self):
         return [sym for sym in self.datasymbols if sym.is_argument]
 
     @property
-    def global_symbols(self):
+    def import_symbols(self):
```
Comment (arporter) [inline src/psyclone/psyir/symbols/symboltable.py:906]: Seeing this in use, I think `imported_symbols` would be a bit nicer.

File: src/psyclone/psyir/symbols/symboltable.py [hunk 10]
```diff
@@ -971,34 +971,35 @@ def data_arguments(self):
...
+                "".format(type(imported_var).__name__))
 
-        if not globalvar.is_global:
+        if not imported_var.is_import:
             raise TypeError(
-                "The globalvar argument of SymbolTable.copy_external_"
-                "global method should have a GlobalInterface interface, "
-                "but found '{0}'.".format(type(globalvar.interface).__name__))
+                "The imported_var argument of SymbolTable.copy_external_"
+                "import method should have a ImportInterface interface, "
```
Comment (arporter) [inline src/psyclone/psyir/symbols/symboltable.py:998]: "...have an ImportInterface..."

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 11]
```diff
@@ -719,7 +719,7 @@ def test_gen_decls_nested_scope(fortran_writer):
 
 def test_gen_decls_routine(fortran_writer):
     '''Test that the gen_decls method raises an exception if the interface
-    of a routine symbol is not a GlobalInterface, unless there's a wildcard
+    of a routine symbol is not a ImportInterface, unless there's a wildcard
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py:722]: "not an"

File: src/psyclone/tests/psyir/symbols/symboltable_test.py [hunk 12]
```diff
@@ -1113,30 +1113,30 @@ def test_unresolved():
...
-    assert "The globalvar argument of SymbolTable.copy_external_global " \
-        "method should have a GlobalInterface interface, but found " \
+        symtab.copy_external_import(DataSymbol("var1", REAL_TYPE))
+    assert "The imported_var argument of SymbolTable.copy_external_import " \
+        "method should have a ImportInterface interface, but found " \
         "'LocalInterface'." \
         in str(error.value)
 
-    # Copy a globalvar
+    # Copy a imported_var
```
Comment (arporter) [inline src/psyclone/tests/psyir/symbols/symboltable_test.py:1135]: s/a/an/

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good Sergi - the names definitely make a lot more sense now :-)
All tests and examples are OK with compilation.
Changes are pycodestyle and pylint clean.
Just a few typos to fix and I have a small query about the precise name of one of the methods.

File: src/psyclone/transformations.py [hunk 13]
```diff
@@ -3306,8 +3306,8 @@ def validate(self, node, options=None):
 
     def apply(self, node, options=None):
         '''
-        Convert the global variables used inside the kernel into arguments and
-        modify the InvokeSchedule to pass the same global variables to the
+        Convert the importe variables used inside the kernel into arguments and
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/transformations.py:3309]: Fixed

File: src/psyclone/psyGen.py (hunk 1 above)
Comment (sergisiso) [reply src/psyclone/psyGen.py:1025]: Done

File: src/psyclone/psyir/backend/fortran.py (hunk 2 above)
Comment (sergisiso) [reply src/psyclone/psyir/backend/fortran.py:683]: Done

File: src/psyclone/psyir/backend/fortran.py (hunk 3 above)
Comment (sergisiso) [reply src/psyclone/psyir/backend/fortran.py:709]: Done

File: src/psyclone/psyir/symbols/symbol.py (hunk 5 above)
Comment (sergisiso) [reply src/psyclone/psyir/symbols/symbol.py:103]: Yes, done

File: src/psyclone/psyir/symbols/symbol.py (hunk 6 above)
Comment (sergisiso) [reply src/psyclone/psyir/symbols/symbol.py:311]: Done

File: src/psyclone/psyir/symbols/symbol.py (hunk 7 above)
Comment (sergisiso) [reply src/psyclone/psyir/symbols/symbol.py:343]: Done

File: src/psyclone/psyir/symbols/symbol.py (hunk 8 above)
Comment (sergisiso) [reply src/psyclone/psyir/symbols/symbol.py:432]: Done

File: src/psyclone/psyir/symbols/symboltable.py (hunk 10 above)
Comment (sergisiso) [reply src/psyclone/psyir/symbols/symboltable.py:998]: Done

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 11 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/fortran_test.py:722]: Done

File: src/psyclone/tests/psyir/symbols/symboltable_test.py (hunk 12 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/symbols/symboltable_test.py:1135]: Done

File: src/psyclone/psyir/nodes/psy_data_node.py (hunk 4 above)
Comment (sergisiso) [reply src/psyclone/psyir/nodes/psy_data_node.py:140]: I agree `imported_symbols` feels more natural. Changed.

File: src/psyclone/psyir/symbols/symboltable.py (hunk 9 above)
Comment (sergisiso) [reply src/psyclone/psyir/symbols/symboltable.py:906]: Done

**Comment by sergisiso** [comment]
@arporter Addressed the comments and brought it to master, this is ready for next review.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Update user guide builds fine and looks good.
All tests and examples are OK with compilation.
Ref. guide builds without new warnings.
Issue no. is not referenced in code, tests, readme's, notebooks or docs.

**Pull request merged**